### PR TITLE
sea3d new examples and revision

### PR DESCRIPTION
**Pull request by sunag**
Hi @mrdoob .

I have a few more examples, I hope you enjoy... morph seems that still is not working in r72dev?

```
console.warn( 'TOFIX: material.program is undefined' );
```

Thks

### CODE REVIEW

File: examples/js/loaders/sea3d/SEA3DLoader.js [hunk 1]
```diff
@@ -25,12 +25,12 @@ THREE.ShaderLib.replaceCode = function(src, target, replace) {
 THREE.ShaderLib['phong'].fragmentShader_StandardMaterial = 
 	THREE.ShaderLib.replaceCode( THREE.ShaderLib['phong'].fragmentShader, [
```
Comment (WestLangley) [inline examples/js/loaders/sea3d/SEA3DLoader.js]: I did not realize this was in your previous version, too.

Why do you feel it is necessary to modify the `Phong` shader code in the sea3d loader?

I notice you set `material.emissive.set( 0.2, 0.2, 0.2 )` (or something similar) on almost all your models.

You appear to want `emissive` to be modulated by both `material.color` and `material.map`.

In the `Phong` shader, `material.emissive` is simply additive to the reflected color.

**Comment by WestLangley** [comment]
I like `webgl_loader_sea3d_hierarchy`. : - )

Consider using `OrbitControls` instead, so your scenes remain right-side-up. If you do so, you should not need to call `controls.update()` in the render loop unless there is camera animation.

**Comment by sunag** [comment]
Hi @WestLangley 

'OrbitControls' looks much better. Thanks : ) 

`Emissive` has a different result of `AmbienteColor`. Many people who use 3ds Max and Sea3D Studio like of Ambient Color and it seems that he is no longer available. So I made this hack to not need create a custom shader for it.

**Comment by sunag** [comment]
I almost forgot... It is possible load without the hack.
```
loader = new THREE.SEA3D( {
	autoPlay : true, // Auto play animations
	container : scene, // Container to add models
	parser : THREE.SEA3D.AUTO, // Auto choose THREE.BufferGeometry and THREE.Geometry
	multiplier : 1, // Light multiplier
	ambientColor : false // emissiveToAmbientColor, true is default
} );
```
Creating manually a material with emissive to ambient color.
```
var mat = new THREE.MeshPhongMaterial();
mat.emissiveToAmbientColor = true; // false is default
```

File: examples/webgl_loader_sea3d.html [hunk 2]
```diff
@@ -72,7 +72,7 @@
 
 				autoPlay : true, // Auto play animations
 				container : scene, // Container to add models
-				parser : THREE.SEA3D.DEFAULT, // THREE.SEA3D.BUFFER to THREE.GeometryBuffer
+				parser : THREE.SEA3D.AUTO, // Auto choose THREE.BufferGeometry and THREE.Geometry
```
Comment (mrdoob) [inline examples/webgl_loader_sea3d.html]: I would suggest not producing `THREE.Geometry` at all (thus removing all the code for it). And let the user convert on their code if needed.

**Comment by mrdoob** [comment]
> I have a few more examples, I hope you enjoy... morph seems that still is not working in r72dev?

Yes, still working on that. I'm hoping to be able to spend some time tomorrow 😊

**Comment by mrdoob** [comment]
Thanks!

**Comment by WestLangley** [comment]
@mrdoob @dubejf @tschw etc..

What are the implications/consequences of the first 400 lines of `SEA3DLoader.js`?

**Comment by WestLangley** [comment]
> Emissive has a different result of AmbientColor. Many people who use 3ds Max and Sea3D Studio like of Ambient Color and it seems that he is no longer available. So I made this hack to not need create a custom shader for it.

`material.ambient` was removed from the interface. In the model, it was hardwired to match `material.color`.

I see that you are adding `emissive` to most materials to accommodate for something you do not like. You should be able to get the same effect by increasing `AmbientLight`.

Your hack to the `Phong` shader is not appropriate. We need to find another solution. `emissive` should not be modulated by the diffuse color and diffuse map.

**Comment by dubejf** [comment]
> What are the implications/consequences of the first 400 lines of SEA3DLoader.js?

You mean extending `THREE` objects? Are all those extensions that really necessary?

I guess it's okay for application code to clobber the `THREE` objects, but ideally, it would be a lot cleaner to extend the library with new classes (e.g. `Sea3DSkinnedMesh` extends `THREE.SkinnedMesh`).

**Comment by sunag** [comment]
> material.ambient was removed from the interface. In the model, it was hardwired to match material.color.

At least in my tests in r71 mat.color took another result, like a * b

> I see that you are adding emissive to most materials to accommodate for something you do not like. You should be able to get the same effect by increasing AmbientLight.

The value .2 can vary from material to material. I can not make a standard value for this at the time.

> Your hack to the Phong shader is not appropriate. We need to find another solution. emissive should not be modulated by the diffuse color and diffuse map.

Also I do not think appropriate, I had talked about changing the way of how the fragment shader is built for easy extension without having create a custom shader for this, like a node material https://github.com/mrdoob/three.js/issues/5673 for the LayeredTexture that I want implement. Away3D make this architecture very well, I possibly do something like node material for the next updates of sea3d.

video demo of sea3d LayeredTexture.
https://www.youtube.com/watch?v=lQ-VIni5Q68

> You mean extending THREE objects? Are all those extensions that really necessary?
I guess it's okay for application code to clobber the THREE objects, but ideally, it would be a lot cleaner to extend the library with new classes (e.g. Sea3DSkinnedMesh extends THREE.SkinnedMesh).

Yes has several extensions, all with a purpose. Sea3DSkinnedMesh seems like a good idea, just do not know if I could with THREE.Object3D for relative keyframe animation and Morph+Skin also be a problem with the interface.

I will do everything possible to improve this, possibly I make a separate version for this extension.

Cheers

**Comment by WestLangley** [comment]
@sunag `material.emissive` is used for objects that glow. You are using it for another purpose.

Are you hacking the shader code because you want `material.ambient` to be different from `material.color` -- and in r.71 they are hardwired to be the same?

**Comment by sunag** [comment]
> @sunag material.emissive is used for objects that glow. You are using it for another purpose.

I converted `emissive` to `ambient` because I have not found another method to do this without custom shader. The best way for me here is create a node material even for custom layers in albedo output without edit shader code. I can do this in the next sea3d updates

**Comment by mrdoob** [comment]
> morph seems that still is not working in r72dev?

I think it should be working fine now.

**Comment by sunag** [comment]
> I think it should be working fine now.

Thanks! Works great.

**Pull request merged**